cmd/snap-confine/snap-confine.apparmor.in

```diff
@@ -550,22 +552,33 @@
         /tmp/.snap/{,**} rw,
         # Allow mounting/unmounting any part of $SNAP over to a temporary place
         # in /tmp/.snap/ during the preparation of a writable mimic.
-        mount options=(bind, rw) /snap/*/** -> /tmp/.snap/**,
+        mount options=(bind, rw) /*/** -> /tmp/.snap/**,
         # Allow mounting tmpfs over the original $SNAP/** directory.
-        mount fstype=tmpfs options=(rw) tmpfs -> /snap/*/**,
+        mount fstype=tmpfs options=(rw) tmpfs -> /*/**,
         # Allow bind mounting anything from the temporary place in /tmp/.snap/
         # back to $SNAP/** (to re-construct the data that was there before).
-        mount options=(bind, rw) /tmp/.snap/** -> /snap/*/**,
+        mount options=(bind, rw) /tmp/.snap/** -> /*/**,
         # Allow unmounting the temporary directory in /tmp once it is no longer
         # necessary.
         umount /tmp/.snap/**,
 
-        # Allow creating missing directories under $SNAP once the writable
-        # mimic is ready.
+        # Allow creating missing directories anywhere under the root directory
+        # (but not in the root directory itself) where they need to be created
+        # as a mount point for layouts or for content sharing. This is a
+        # superset of other cases so they are removed
         / r,
-        /snap/ r,
-        /snap/{,**} r,
-        /snap/*/** w,
+        /** r,
+        /*/** w,
+
+        # Allow layouts to bind mount *from* $SNAP, $SNAP_DATA and $SNAP_COMMON
+        # *to* anywhere under the root directory. This is safe because the
+        # mounts happen inside an isolated mount namespace (but see below).
+        mount options=(bind) /snap/*/** -> /*/**,
+        mount options=(bind) /var/snap/*/** -> /*/**,
+        # TODO: allow that ^ for symlink as well
+        # As an exception, don't allow bind mounts to /media which has special
+        # sharing and propagates mount events outside of the snap namespace.
+        audit deny mount -> /media,
```

Comment:
This may cause problems with the udisks2 slot implementations, no? I guess they will not be able to use layouts? This should be documented here I think.

Reply:
No, it's just for the snap-update-ns, the actual snap can still do this (via the udisks interface)

Reply:
Oh right, duh.